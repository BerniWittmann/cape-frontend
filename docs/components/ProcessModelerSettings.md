# ProcessModelerSettings

The Settings Tag to edit the settings for the Process Modeler

