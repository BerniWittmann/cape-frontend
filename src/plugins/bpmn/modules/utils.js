import {
  NO_CONNECTIONS_DISALLOWED_OBJECTS,
  END_EVENT_TYPE,
  START_EVENT_TYPE,
  MULTIPLE_SEQUENCE_FLOW_ALLOWED_OBJECT_TYPES,
  CONNECTION_RULE_ENFORCED_ELEMENTS,
  SEQUENCE_FLOW_TYPE
} from './constants'

export function shapeIsType(shape, type) {
  if (!shape || !shape.businessObject) return false
  return shape.businessObject.$type === type
}

export function shapeHasType(shape, types) {
  if (!shape || !shape.businessObject) return false
  return types.includes(shape.businessObject.$type)
}

export function getElementsByType(elementRegistry, type) {
  return elementRegistry.filter(el => shapeIsType(el, type))
}

export function hasUnconnectedActivities(elementRegistry) {
  const unconnectedElements = elementRegistry.filter(el => {
    if (!shapeHasType(el, NO_CONNECTIONS_DISALLOWED_OBJECTS)) return false
    if (el.type === 'label') return false
    const hasIncomingConnections = el.incoming.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length > 0
    const hasOutgoingConnections = el.outgoing.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length > 0
    if (shapeIsType(el, END_EVENT_TYPE) && hasIncomingConnections) return false
    if (shapeIsType(el, START_EVENT_TYPE) && hasOutgoingConnections) return false
    return !(hasIncomingConnections && hasOutgoingConnections)
  })
  return unconnectedElements.length > 0
}

export function hasExceedingConnectionCountElements(elementRegistry) {
  const invalidConnectedElements = elementRegistry.filter(el => {
    if (shapeHasType(el, MULTIPLE_SEQUENCE_FLOW_ALLOWED_OBJECT_TYPES)) return false
    if (!shapeHasType(el, CONNECTION_RULE_ENFORCED_ELEMENTS)) return false
    const incomingConnections = el.incoming.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length
    const outgoingConnections = el.outgoing.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length
    if (shapeIsType(el, END_EVENT_TYPE) && incomingConnections === 1) return false
    if (shapeIsType(el, START_EVENT_TYPE) && outgoingConnections === 1) return false
    return !(incomingConnections === 1 && outgoingConnections === 1)
  })
  return invalidConnectedElements.length > 0
}

export function getStartEventCount(elementRegistry) {
  const startEvents = elementRegistry.filter(el => {
    return shapeIsType(el, START_EVENT_TYPE)
  })
  return startEvents.length
}

export function getEndEventCount(elementRegistry) {
  const endEvents = elementRegistry.filter(el => {
    return shapeIsType(el, END_EVENT_TYPE)
  })
  return endEvents.length
}
