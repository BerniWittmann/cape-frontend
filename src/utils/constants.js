export const DATE_TIME_FORMAT = 'DD.MM.YYYY HH:mm'
export const DEFAULT_TAG_COLOR = '#6CC9E8'

export const PROCESS_VALIDATION_ERRORS = {
  UNCONNECTED_ENTITIES: 'unconnected_entities',
  NO_START_EVENT: 'no_start_event',
  TOO_MANY_START_EVENTS: 'too_many_start_events',
  NO_END_EVENT: 'no_end_event',
  TOO_MANY_END_EVENTS: 'too_many_end_events',
  ELEMENT_CONNECTIONS_EXCEEDED: 'element_connections_exceeded'
}
