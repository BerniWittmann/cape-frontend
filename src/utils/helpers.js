export function convertHexToRgba(color, opacity = 1) {
  const c = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(color)
  return `rgba(${parseInt(c[1], 16)}, ${parseInt(c[2], 16)}, ${parseInt(c[3], 16)}, ${opacity})`
}

export function getCookie(name) {
  const value = '; ' + document.cookie
  const parts = value.split('; ' + name + '=')
  if (parts.length === 2) return parts.pop().split(';').shift()
}

export function hasProcessModelerRulesEnabled() {
  const cookie = getCookie('enableProcessValidation')
  return cookie ? cookie === 'true' : true
}

export function removeByID(arr, id) {
  return arr.filter(o => o.id !== id)
}

export function updateAndSetActive(store, data, Model, moduleName) {
  data = new Model(data)
  store.dispatch(`${moduleName}/update`, data).then(() => {
    store.dispatch(`${moduleName}/setActive`, data)
  })
}
