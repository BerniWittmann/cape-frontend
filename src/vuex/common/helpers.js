import Vue from 'vue'

function pushToChildren(tagMap, id, data) {
  const tag = tagMap.get(id)
  if (!tag) return
  tag.children.push(data)
  tagMap.set(id, tag)
}

export function update(state, key, activeKey, data) {
  state[key] = [...state[key].map((p) => p.id === data.id ? { ...data } : p)]
  if (state[activeKey] && state[activeKey].id === data.id) {
    state[activeKey] = { ...data }
  }
}

export function setActive(state, key, activeKey, data) {
  if (!data || !data.id) {
    state[activeKey] = undefined
  } else {
    state[activeKey] = state[key].find((p) => p.id === data.id)
  }
}

export function getGraphNodes(objects, type, routeName, routeID) {
  return objects.map((o) => ({
    id: o.id,
    name: o.name,
    type: type,
    route: {
      name: routeName,
      params: {
        [routeID]: o.id
      }
    }
  }))
}

export function getByTags(objects, tags) {
  const tagMap = new Map()
  tagMap.set('untagged', {
    value: '0',
    label: Vue.i18n.t('context_situation.untagged'),
    children: []
  })
  tags.forEach(tag => {
    tagMap.set(tag.id, {
      value: tag.id,
      label: tag.name,
      children: []
    })
  })
  objects.forEach((obj) => {
    const objData = {
      value: obj,
      key: obj.id,
      label: obj.name
    }
    if (obj.tags && obj.tags.length > 0) {
      obj.tags.forEach((tag) => {
        pushToChildren(tagMap, tag.id, objData)
      })
    } else {
      pushToChildren(tagMap, 'untagged', objData)
    }
  })
  return [...tagMap.values()]
}
