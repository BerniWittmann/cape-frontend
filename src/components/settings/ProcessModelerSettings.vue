<template>
  <el-row type="flex" justify="center">
    <el-col :span="8">
      <el-checkbox v-model="enableValidation" @change="setCookie" :label="$t('settings.process_modeler.enable_validation')" border></el-checkbox>
      <p><small>{{ $t('settings.process_modeler.enable_validation_help_text') }}</small></p>
    </el-col>
  </el-row>
</template>

<script>
/* ============
 * Process Modeler Settings Component
 * ============
 *
 * A component to edit the settings for the Process Modeler
 */

import { hasProcessModelerRulesEnabled } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * The Settings Tag to edit the settings for the Process Modeler
 */
export default {
  name: 'ProcessModelerSettings',
  data() {
    return {
      enableValidation: false
    }
  },

  methods: {
    getCookieValue() {
      this.enableValidation = hasProcessModelerRulesEnabled()
    },

    setCookie(val) {
      document.cookie = 'enableProcessValidation=' + val + '; expires=Fri, 31 Dec 2100 12:00:00 UTC; path=/'
      this.getCookieValue()
    }
  },

  beforeMount() {
    this.getCookieValue()
  }
}

</script>

<style scoped lang="scss">
</style>
