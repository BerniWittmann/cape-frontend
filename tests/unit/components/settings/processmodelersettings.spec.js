import { mount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'
import { Checkbox } from 'element-ui'

import ProcessModelerSettings from '@/components/settings/ProcessModelerSettings.vue'

import * as helpers from '@/utils/helpers'

describe('Components', () => {
  describe('ProcessModelerSettings', () => {
    let cmp

    helpers.hasProcessModelerRulesEnabled = jest.fn().mockReturnValue(true)

    beforeEach(() => {
      render()
    })

    function render() {
      cmp = mount(ProcessModelerSettings, {
        i18n,
        stubs: {
          Checkbox
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    describe('it can toggle the custom Rules', () => {
      describe('loads the existing state on render', () => {
        it('loads the active state', () => {
          helpers.hasProcessModelerRulesEnabled = jest.fn().mockReturnValue(true)
          render()

          expect(cmp.html()).toMatchSnapshot()
          expect(cmp.find('.el-checkbox__input').classes('is-checked')).toBeTruthy()
        })

        it('loads the inactive state', () => {
          helpers.hasProcessModelerRulesEnabled = jest.fn().mockReturnValue(false)
          render()

          expect(cmp.html()).toMatchSnapshot()
          expect(cmp.find('.el-checkbox__input').classes('is-checked')).toBeFalsy()
        })
      })

      describe('it writes the state to a cookie', () => {
        it('can activate the state', () => {
          helpers.hasProcessModelerRulesEnabled = jest.fn().mockReturnValue(false)
          render()
          cmp.find('input').setChecked(true)

          expect(document.cookie).toEqual('enableProcessValidation=false')
        })

        it('can deactivate the state', () => {
          helpers.hasProcessModelerRulesEnabled = jest.fn().mockReturnValue(true)
          render()
          cmp.find('input').setChecked(false)

          expect(document.cookie).toEqual('enableProcessValidation=true')
        })
      })
    })
  })
})
