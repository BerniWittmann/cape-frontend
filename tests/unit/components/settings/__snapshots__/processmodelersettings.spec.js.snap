// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ProcessModelerSettings it can toggle the custom Rules loads the existing state on render loads the active state 1`] = `
<div class="el-row is-justify-center el-row--flex">
  <div class="el-col el-col-8"><label role="checkbox" aria-checked="true" class="el-checkbox is-bordered is-checked"><span aria-checked="mixed" class="el-checkbox__input is-checked"><span class="el-checkbox__inner"></span><input type="checkbox" aria-hidden="true" class="el-checkbox__original" value="settings.process_modeler.enable_validation"></span><span class="el-checkbox__label">settings.process_modeler.enable_validation</span></label>
    <p><small>settings.process_modeler.enable_validation_help_text</small></p>
  </div>
</div>
`;

exports[`Components ProcessModelerSettings it can toggle the custom Rules loads the existing state on render loads the inactive state 1`] = `
<div class="el-row is-justify-center el-row--flex">
  <div class="el-col el-col-8"><label role="checkbox" class="el-checkbox is-bordered"><span aria-checked="mixed" class="el-checkbox__input"><span class="el-checkbox__inner"></span><input type="checkbox" aria-hidden="true" class="el-checkbox__original" value="settings.process_modeler.enable_validation"></span><span class="el-checkbox__label">settings.process_modeler.enable_validation</span></label>
    <p><small>settings.process_modeler.enable_validation_help_text</small></p>
  </div>
</div>
`;

exports[`Components ProcessModelerSettings renders 1`] = `
<div class="el-row is-justify-center el-row--flex">
  <div class="el-col el-col-8"><label role="checkbox" aria-checked="true" class="el-checkbox is-bordered is-checked"><span aria-checked="mixed" class="el-checkbox__input is-checked"><span class="el-checkbox__inner"></span><input type="checkbox" aria-hidden="true" class="el-checkbox__original" value="settings.process_modeler.enable_validation"></span><span class="el-checkbox__label">settings.process_modeler.enable_validation</span></label>
    <p><small>settings.process_modeler.enable_validation_help_text</small></p>
  </div>
</div>
`;
